soweego/linker/feature_extraction.py

```diff
@@ -177,6 +178,101 @@ def exact_apply(pair):
+
+    name = "DateCompare"
+    description = "Compares the date attribute of record pairs."
+
+    def __init__(self,
+                 left_on,
+                 right_on,
+                 missing_value=0.0,
+                 label=None):
+        super(DateCompare, self).__init__(left_on, right_on, label=label)
+
+        self.missing_value = missing_value
+
+    def _compute_vectorized(self, source_column, target_column):
+
+        # we zip together the source column and the target column so that
+        # they're easier to process
+        concatenated = pd.Series(list(zip(source_column, target_column)))
+
+        def check_date_equality(pair: Tuple[Union[pd.Period, List[pd.Period]], pd.Period]):
+            """
+            Compares a target pd.Period with the source pd.Periods which represent either
+            a birth or death date. The source date can be a list of possible dates.
+
+            Returns the most optimistic match
+            """
+
+            s_item, t_item = pair
+
+            # if t_item is NaT then we can't compare, and we skip this pair
+            if pd.isna(t_item):
+                return 0
+
+            # convert `s_item` to a list if it isn't already
+            if not isinstance(s_item, (list, tuple)):
+                s_item = [s_item]
+
+            # precisions listed from least to most precise, as defined here:
+            # http://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#dateoffset-objects
+            precisions = [
+                'A-DEC',  # we know only the year
+                'M',  # we know up to the month
+                'D',  # up to the day
+                'H',  # up to the hour
+                'T',  # up to the minute
+                'S',  # up to the second
+                'U',  # up to the microsecond
+                'N',  # up to the nanosecond
+            ]
+
+            # will help us to keep track of the best score
+            best = 0
+
+            for s_date in s_item:
+
+                # if the current s_date is NaT then we can't compare, so we skip it
+                if pd.isna(s_date):
+                    continue
+
+                # get precision number for both dates
+                s_precision = precisions.index(s_date.freq.name)
+                t_precision = precisions.index(t_item.freq.name)
+
+                # we choose to compare on the lowest precision
+                # since it's the maximum precision on which both
+                # dates can be safely compared
+                lowest_prec = min(s_precision, t_precision)
+
+                # the result for the current `s_date`
+                c_r = 0
+
+                for min_required_prec, d_attr in enumerate(['year', 'month', 'day', 'hour', 'minute', 'second']):
+                    if lowest_prec >= min_required_prec and getattr(s_date, d_attr) == getattr(t_item, d_attr):
+                        c_r += 1
+                    else:
+                        break
```

Comment:
Yes, it might be a bit unclear. I added a just comment explaining this in bef93d33cf54193911fb61c868edab9173991d7d
Anyway, currently we consider from lowest to highest precision. If a lowest precision attribute (ie, year) doesn't match then we say that the dates don't match at all (we don't check if higher precision attributes match).

Do you think we should check all possible parts of the date?

Reply:
Thanks for the explanation, now it's clear.
I think your solution is nicely based on common sense: if years are different, then the date pair is not the same.
However, my gut feeling is that data is often messy, and dates can come in whatever format.

We should try to do our best to normalize and compare them.

I suggest to indeed check all possible parts, but to assign different weights:
- if the years match, then maximum weight;
- if the days mach, then medium weight;
- if the months match, then low weight.

And so on with other time units.

Reply:
Said that, I think it is out of scope for this pull request, so I'm resolving this conversation and approving the PR.